src/phoenix/server/app.py

```diff
@@ -356,29 +380,33 @@ def create_app(
     tracer_provider = None
     strawberry_extensions = schema.get_extensions()
     if server_instrumentation_is_enabled():
-        from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
-        from opentelemetry.trace import TracerProvider
-        from strawberry.extensions.tracing import OpenTelemetryExtension
+        if engine := db.engine:
+            from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
+            from opentelemetry.trace import TracerProvider
+            from strawberry.extensions.tracing import OpenTelemetryExtension
+
+            tracer_provider = initialize_opentelemetry_tracer_provider()
+            SQLAlchemyInstrumentor().instrument(
+                engine=engine.sync_engine,
+                tracer_provider=tracer_provider,
+            )
+            clean_ups.append(SQLAlchemyInstrumentor().uninstrument)
+            if TYPE_CHECKING:
+                # Type-check the class before monkey-patching its private attribute.
+                assert OpenTelemetryExtension._tracer
+
+            class _OpenTelemetryExtension(OpenTelemetryExtension):
+                def __init__(self, *args: Any, **kwargs: Any) -> None:
+                    super().__init__(*args, **kwargs)
+                    # Monkey-patch its private tracer to eliminate usage of the global
+                    # TracerProvider, which in a notebook setting could be the one
+                    # used by OpenInference.
+                    self._tracer = cast(TracerProvider, tracer_provider).get_tracer("strawberry")
+
+            strawberry_extensions.append(_OpenTelemetryExtension)
+        else:
+            logger.info("Instrumenting is not supported in tests.")
```

Comment:
I agree, but because we're passing around a session maker instead of an engine directly around the application passing in an actual engine here created complications for our test fixtures. I think probably the "right" refactor is to stop passing around the `session` generator object and simply pass an engine around instead

Reply:
I don't necessarily agree with the code not being deferential 100% though, I think a lot of patterns that are useful are explicitly used because they make testing easier

Reply:
Still, it's not clear why this accommodation is necessary since the tests can just coerce the instrumentation flag to be false, and the issue is moot.

Reply:
Coercing the instrumentation flag to false encodes a "hidden" bug IMO, if someone writes a test for database instrumentation later, the failure will not be obvious. I agree right now it's moot, but give me a little bit to try and find something a little better

Reply:
Since tests should be run in a pristine environment, by "coerce" I'm just referring to the situation where the environment is leaky and have environment variable contamination, e.g. running the debugger on local IDE.